tests/filecheck/transforms/reuse-memref-space.mlir

```diff
@@ -0,0 +1,319 @@
+//CHECK-NEXT:     %5 = "memref.dim"(%arg0, %1) : (memref<?x?xi8, "L3">, index) -> index
+//CHECK-NEXT:     %6 = memref.alloc(%4, %5) {"alignment" = 64 : i64} : memref<?x?xi8, #tsl.tsl<[?, 8] -> (?, 8), [?, 8] -> (256, 1)>, "L1">
+//CHECK-NEXT:     scf.for %arg2 = %3 to %0 step %1 {
+//CHECK-NEXT:     }
+//CHECK-NEXT:     func.return
+//CHECK-NEXT:   }
+//CHECK-NEXT: }
+// -----
+
+builtin.module {
+  func.func @streamer_matmul(%arg0 : memref<?x?xi8, "L3">, %arg1 : memref<?x?xi32, "L3">) {
+    %0 = arith.constant 8 : index
+    %1 = arith.constant 1 : index
+    %2 = arith.constant 0 : i32
+    %3 = arith.constant 0 : index
+    scf.for %arg2 = %3 to %0 step %1{
+        %4 = "memref.dim"(%arg0, %3) :(memref<?x?xi8, "L3">, index) -> index
+        %5 = "memref.dim"(%arg0, %1) :(memref<?x?xi8, "L3">, index) -> index
+        %6 = memref.alloc(%4, %5) {"alignment" = 64 : i64} : memref<?x?xi8, #tsl.tsl<[?, 8] -> (?, 8), [?, 8] -> (256, 1)>, "L1">
+        }
+    func.return
+  }
+}
+
+// Check if the memref.alloc is elevated outside together with dim operations inside loop
+//CHECK: builtin.module {
+//CHECK-NEXT:   func.func @streamer_matmul(%arg0 : memref<?x?xi8, "L3">, %arg1 : memref<?x?xi32, "L3">) {
+//CHECK-NEXT:     %0 = arith.constant 8 : index
+//CHECK-NEXT:     %1 = arith.constant 1 : index
+//CHECK-NEXT:     %2 = arith.constant 0 : i32
+//CHECK-NEXT:     %3 = arith.constant 0 : index
+//CHECK-NEXT:     %4 = arith.constant 0 : index
+//CHECK-NEXT:     %5 = "memref.dim"(%arg0, %4) : (memref<?x?xi8, "L3">, index) -> index
+//CHECK-NEXT:     %6 = arith.constant 1 : index
+//CHECK-NEXT:     %7 = "memref.dim"(%arg0, %6) : (memref<?x?xi8, "L3">, index) -> index
+//CHECK-NEXT:     %8 = memref.alloc(%5, %7) {"alignment" = 64 : i64} : memref<?x?xi8, #tsl.tsl<[?, 8] -> (?, 8), [?, 8] -> (256, 1)>, "L1">
+//CHECK-NEXT:     scf.for %arg2 = %3 to %0 step %1 {
+//CHECK-NEXT:       %9 = "memref.dim"(%arg0, %3) : (memref<?x?xi8, "L3">, index) -> index
+//CHECK-NEXT:       %10 = "memref.dim"(%arg0, %1) : (memref<?x?xi8, "L3">, index) -> index
+//CHECK-NEXT:     }
+//CHECK-NEXT:     func.return
+//CHECK-NEXT:   }
+//CHECK-NEXT: }
+// -----
+
+builtin.module {
+  func.func @streamer_matmul(%arg0 : memref<?x?xi8, "L3">, %arg1 : memref<?x?xi32, "L3">) {
+    %0 = arith.constant 8 : index
+    %1 = arith.constant 1 : index
+    %2 = arith.constant 0 : i32
+    %3 = arith.constant 0 : index
+    scf.for %arg2 = %3 to %0 step %1{
+        %4 = "affine.min"(%1, %0) <{"map" = affine_map<(d0)[s0] -> (8, ((d0 * -1) + s0))>}> : (index, index) -> index
+        %5 = "memref.dim"(%arg0, %1) :(memref<?x?xi8, "L3">, index) -> index
+        %6 = memref.alloc(%4, %5) {"alignment" = 64 : i64} : memref<?x?xi8, #tsl.tsl<[?, 8] -> (?, 8), [?, 8] -> (256, 1)>, "L1">
+        }
+    func.return
+  }
+}
+
+// Check if the memref.alloc is moved out of the loop when affine.min is inside loop
+// so that a new arith.constant is created equal to maximum of affine.min
```

Comment:
~I feel like doing something like "unrolling" your affine min's should not be part of this pass. Can't you just hoist the affine.min outside of the loop as well?~
Okay it is actually defined as a loop carried variable (so you can not just hoist it out), but that is not the case in this test?